02_Development/Display/i2c_interface.py

```diff
@@ -0,0 +1,221 @@
+            The device register you're interested in.
+        to_write: int
+            The data you'd like to write to that register.
+        """
+        pass
+
+    @abstractmethod
+    def write_data(self, byte):
+        """Writes a single byte to the the device address.
+
+        Parameters
+        ----------
+        byte: int
+            The byte you'd like to write to the device address.
+        """
+
+
+try:
+
+    import board
+    import busio
+
+    class I2CInterface(busio.I2C, I2CInterfaceBase):
+
+        def __init__(self, address, dump_communication=False):
+            self._i2c_address = address
+            self._dump_communication = dump_communication
+            busio.I2C(board.SCL, board.SDA)
+            while not self.try_lock():
+                pass
+
+        def close(self):
+            self.unlock()
+            self.deinit()
+
+        def find_device(self, timeout=5):
+            start_time = time.time()
+            while not self._i2c_address in self.scan():
+                if time.time() - start_time > timeout:
+                    raise DeviceNotFoundError("Could not find I2C "
+                                              "device at address "
+                                              f"{self._i2c_address}")
+
+        def read_register(self, register, number_of_bytes=1):
+            return self._read(register=register,
+                              number_of_bytes=number_of_bytes)
+
+        def read_data(self, number_of_bytes=1):
+            return self._read(number_of_bytes=number_of_bytes)
+
+        def write_register(self, register, to_write):
+            self.writeto(self._i2c_address, bytes([register, to_write]))
+
+            if self._dump_communication:
+                print(f"{1000*time.time():.4f} "
+                      "TX -> 0x" + bytes([register, to_write]).hex())
+
+        def write_data(self, data):
+            self.writeto(self._i2c_address, bytes([data]))
+
+            if self._dump_communication:
+                print(f"{1000*time.time():.4f} TX -> 0x"
+                      + bytes([data]).hex())
+
+        def _read(self, register=None, number_of_bytes=1):
+            if number_of_bytes < 1:
+                raise ValueError("Cannot read fewer than 1 byte.")
+
+            data = bytearray(number_of_bytes)
+            if register is not None:
+                self.writeto(self._i2c_address, bytes([register]))
+            self.readfrom_into(self._i2c_address, data, end=number_of_bytes)
+
+            if self._dump_communication:
+                if register is not None:
+                    print(f"{1000*time.time():.4f} TX -> 0x"
+                          + bytes([register]).hex())
+                print(f"{1000*time.time():.4f} RX <- {data:#02X}")
+
+            if number_of_bytes == 1:
+                return int(data.hex(), 16)
+            else:
+                return tuple(int(byte.hex(), 16) for byte in data)
+
+
+except NotImplementedError:
```

Comment:
this seems dangling

Reply:
There's an implementation of `I2CInterface` after that.  Is this another instance of Github being bad?  Because it's definitely in the repo.